### PR TITLE
Auth: Add skip_org_role_sync setting for GrafanaCom

### PR DESCRIPTION
**Pull request by eleijonmarck**
**What is this feature?**
- adds `skip_org_role_sync` setting for auth.grafana_com
- adds setting reading for section
- rename struct field (not the setting) from skipOrgRoleSync in SocialBase to skipOrgRoleSyncBase
- added docs under auth/grafana-com
- added test for userInfo
- [x] TODO: add a table of the configuration for both skipOrgRoleSync and the skipOrgRoleSync for GrafanaCom, what happens when both are set separately and both are set.
- [x] updated whats new https://github.com/grafana/grafana/pull/59614

**Why do we need this feature?**

We have customers asking to set their own roles for a user within the grafana instance without syncing it with grafanacom

**Which issue(s) does this PR fix?**:
https://github.com/grafana/grafana-enterprise/issues/4235

Fixes #

**Special notes for your reviewer**:

This shows the feature in which has been done

https://user-images.githubusercontent.com/7727602/212026994-b34142f4-68d9-4ca3-80cb-51a92b4979c3.mp4

### CODE REVIEW

File: pkg/login/social/social.go [hunk 1]
```diff
@@ -309,7 +310,7 @@ func newSocialBase(name string,
 		autoAssignOrgRole:       autoAssignOrgRole,
 		roleAttributePath:       info.RoleAttributePath,
 		roleAttributeStrict:     info.RoleAttributeStrict,
-		skipOrgRoleSync:         skipOrgRoleSync,
+		skipOrgRoleSyncBase:     skipOrgRoleSync,
```
Comment (alexanderzobnin) [inline pkg/login/social/social.go:313]: Why do we need `skipOrgRoleSyncBase` instead of `skipOrgRoleSync`?

Comment (eleijonmarck) [reply pkg/login/social/social.go:313]: We do not really need it right now. Thought I'd rename to have to separate fields that can be set from the configuration. One in the overall scope for Auth (the one we want to deprecate) and one that is set for the specific OAuth integration.

Does this make sense?

File: packages/grafana-data/src/types/config.ts [hunk 2]
```diff
@@ -220,8 +220,11 @@ export interface GrafanaConfig {
 }
 
 export interface AuthSettings {
+  // deprecation notice
+  // in favor of setting one for each specific auth integration
```
Comment (IevaVasiljeva) [inline packages/grafana-data/src/types/config.ts:224]: How are deprecation notices added for settings? This is an internal file and is not very visible, so I think we'll need to add more than this for customers to be aware of the change. I'd prefer creating a new issue to track the work so that we make sure to do it properly.

File: pkg/login/social/grafana_com_oauth.go [hunk 3]
```diff
@@ -60,12 +63,17 @@ func (s *SocialGrafanaCom) UserInfo(client *http.Client, _ *oauth2.Token) (*Basi
 		return nil, fmt.Errorf("Error getting user info: %s", err)
 	}
 
+	// on login we do not want to display the role from the external provider
```
Comment (IevaVasiljeva) [inline pkg/login/social/grafana_com_oauth.go:66]: I would move this comment one line below (just above the conditional checking `skipOrgRoleSync` value). I'd also reword it to something like 
```suggestion
	// do not sync the role from the external provider
```
to make it clearer that it isn't just about what we display but also about how user role is determined on the backend.

File: pkg/setting/setting.go [hunk 4]
```diff
@@ -1102,6 +1104,7 @@ func (cfg *Cfg) Load(args CommandLineArgs) error {
 	}
 
 	// check old key  name
+	// TODO: move these into the read grafanacom section
```
Comment (IevaVasiljeva) [inline pkg/setting/setting.go:1107]: Nit: either do it or remove the comment.

File: pkg/setting/setting.go [hunk 5]
```diff
@@ -1351,6 +1354,12 @@ func readSecuritySettings(iniFile *ini.File, cfg *Cfg) error {
 	return nil
 }
 
+func readAuthGrafanaComSettings(iniFile *ini.File, cfg *Cfg) (err error) {
```
Comment (IevaVasiljeva) [inline pkg/setting/setting.go:1357]: nit: this function currently never returns an error, so it could just have an empty return to simplify the code.

File: pkg/login/social/social.go [hunk 6]
```diff
@@ -218,6 +218,7 @@ func ProvideService(cfg *setting.Cfg, features *featuremgmt.FeatureManager) *Soc
 				SocialBase:           newSocialBase(name, &config, info, cfg.AutoAssignOrgRole, cfg.OAuthSkipOrgRoleUpdateSync, *features),
 				url:                  cfg.GrafanaComURL,
 				allowedOrganizations: util.SplitString(sec.Key("allowed_organizations").String()),
+				skipOrgRoleSync:      sec.Key("skip_org_role_sync").MustBool(false),
```
Comment (IevaVasiljeva) [inline pkg/login/social/social.go:221]: I think you could just use `cfg.GrafanaComSkipOrgRoleSync` here.

File: pkg/login/social/social.go (hunk 1 above)
Comment (IevaVasiljeva) [reply pkg/login/social/social.go:313]: We already have the option to specify `skip_org_role_sync` under different sections in the configuration. I also don't really see why this change is necessary.

**Review by IevaVasiljeva — COMMENTED** [review]
I've added some comments inline, but also have some larger questions that should be answered before merging this:
* what is the intended behaviour if `skip_org_role_sync` is specified for both GCom and generic OAuth? Currently roles for GCom will only be synced if both of these config options are set to `false`. I think that makes sense, but I wanted to double-check. We should also document this.
* and talking about documentation, we should add docs for this new config option in general (or are you planning to do that in a separate PR?)
* finally, tests - there don't seem to be any for GCom auth, but it would be good to add some (could also be done in a separate PR).

File: packages/grafana-data/src/types/config.ts (hunk 2 above)
Comment (eleijonmarck) [reply packages/grafana-data/src/types/config.ts]: the idea was to add a comment around it. was not meant as a feature, will instead make a new issue where we add the feature 👍

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
+---
+aliases:
+  - ../../../auth/grafana-com/
+description: Grafana Com Authentication
+title: Configure Grafana Com Authentication
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:5]:
```suggestion
title: Configure Grafana Com authentication
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+weight: 500
+---
+
+# Configure Grafana Com authentication
+
+To enable GrafanaCom as your authentication provider. You configure GrafanaCom to generate a client ID and a secret key for you to use.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:11]:
```suggestion
To enable GrafanaCom as your authentication provider, you configure it to generate a client ID and a secret key.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+## Create Grafana Com OAuth keys
+
+To use Grafana Com authentication:
+
+1. Login on to Grafana com
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:17]:
```suggestion
1. Log in to Grafana Com.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+2. Create an OAuth client by visiting your organization and clicking “OAuth Clients”.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:18]:
```suggestion
1. To create an OAuth client, locate your organization and click **OAuth Clients**. 
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+3. Then click “Add OAuth Client Application”.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:19]:
```suggestion
1. Click **Add OAuth Client Application**.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+4. Add the URL of your running instance of Grafana.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:20]:
```suggestion
1. Add the URL of your running Grafana instance.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+5. The next step will give you the client ID and secret.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:21]: i'm not sure what this means. What next step? After they enter the URL, what do they do next? Maybe there's a submit button or a Next button?

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+Below is an example of the configuration:
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:23]:
```suggestion
The following code shows an example configuration:
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+client_secret = eyJ0Ijoib2F1dGgyYyIhlmlkIjoiNzUwYmMzM2MxMGRjMjE6NDh3OWQiLCJ2IjoiZmI1YzVlYmIwYzFmN2ZhYzZmNjIwOGI1NmVkYTRlNWYxMzgwM2NkMiJ9
+scopes = user:email
+allowed_organizations = sampleorganization
+enabled = true
+
+# prevents the sync of org roles from Grafana.com
+skip_org_role_sync = false
+```
+
+## Skip Org Role Sync
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:39]:
```suggestion
## Skip organization role sync
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+To prevent the sync of org roles from Grafana.com, set `skip_org_role_sync` to `true`. This is useful if you want to use Grafana.com as an identity provider but want to manage the org roles in Grafana.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:41]: I'm a little confused on this one. Why is preventing the sync of org roles a good thing? What happens if they are sync'd? I'd think you'd want them to be sync'd.

**Review by chri2547 — COMMENTED** [review]
A few suggestions and comments. Thank you!

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md (hunk 16 above)
Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:41]: syncing roles from their idp's is the default for most customers. Some customers ask for this when they have multiple providers and would want to be able to manage roles within Grafana without them being synced from their IDP

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md (hunk 13 above)
Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md]: @chri2547 nice to point this out, I have made it a bit more clear with this commit: 97ed25bf87

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+To enable GrafanaCom as your authentication provider, you configure it to generate a client ID and a secret key.
+
+## Create Grafana Com OAuth keys
+
+To use Grafana Com authentication:
+
+1. Log in to Grafana Com.
+1. To create an OAuth client, locate your organization and click **OAuth Clients**.
+1. Click **Add OAuth Client Application**.
+1. Add the name, and URL of your running Grafana instance.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:20]:
```suggestion
1. Add the name and URL of your running Grafana instance.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+1. Click **Add OAuth Client**
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:21]:
```suggestion
1. Click **Add OAuth Client**.
```

File: conf/sample.ini [hunk 19]
```diff
@@ -679,7 +680,7 @@
 # At 1 am every day
 ;sync_cron = "0 1 * * *"
 ;active_sync_enabled = true
-
+;
```
Comment (alexanderzobnin) [inline conf/sample.ini:683]: This one looks unnecessary

File: conf/sample.ini [hunk 20]
```diff
@@ -1308,7 +1309,7 @@
 ; enabled = false
 ; root_ca_cert =
 ; client_key =
-; client_cert = 
+; client_cert =
 ; server_name =
 # The address of the socks5 proxy datasources should connect to
-; proxy_address =
\ No newline at end of file
+; proxy_address =
```
Comment (alexanderzobnin) [inline conf/sample.ini:1315]: I'm not sure is it a blank line in the end of the file? Original one does not have it. Could you double check? Sometimes editor adds new line on save.

**Review by alexanderzobnin — DISMISSED** [review]
Looks good, just a few minor things.

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+
+# Configure Grafana Com authentication
+
+To enable GrafanaCom as your authentication provider, you configure it to generate a client ID and a secret key.
+
+## Create Grafana Com OAuth keys
+
+To use Grafana Com authentication:
+
+1. Log in to Grafana Com.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:17]: How does the user do this? what is the URL?

File: conf/defaults.ini [hunk 22]
```diff
@@ -1388,4 +1389,4 @@ client_key =
 client_cert =
 server_name =
 # The address of the socks5 proxy datasources should connect to
-proxy_address =
\ No newline at end of file
+proxy_address =
```
Comment (eleijonmarck) [inline conf/defaults.ini:1392]: this is just formatting

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 23]
```diff
@@ -858,6 +860,21 @@ By default, the users' organization and role is reset on every new login.
 > With Grafana 10, if `oauth_skip_org_role_update_sync` option is set to `false`, users with no mapping will be
 > reset to the default organization role on every login. [See `auto_assign_org_role` option]({{< relref ".#auto_assign_org_role" >}}).
 
+### [auth].<provider> skip_org_role_update_sync
```
Comment (eleijonmarck) [inline docs/sources/setup-grafana/configure-grafana/_index.md:863]: @IevaVasiljeva updated the PR with some documentation as well for reflecting what happens specifically to the org roles with the different settings set.

Hope this makes sense for the user.

Also; the placement of this is meant for now to be alongside the "generic" setting to make it explicit that we are working towards making this the default.

Not sure where to place this otherwise 🤔 , any pointers there?

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 24]
```diff
@@ -858,6 +860,21 @@ By default, the users' organization and role is reset on every new login.
...
+
+If you want to prevent org roles from being synced for a specific OAuth integration, set `skip_org_role_update_sync` to `true`.
+
+The `oauth_skip_org_role_update_sync` are independent of each other. `skip_org_role_sync` is specific to the provider integration, while `oauth_skip_org_role_sync_update` is a generic setting that only applies to `Github`, `Gitlab`, `AzureAD`, `Generic OAuth` integrations. We are deprecating the generic setting in favor of the provider specific settings.
```
Comment (Jguer) [inline docs/sources/setup-grafana/configure-grafana/_index.md:867]: > The `oauth_skip_org_role_update_sync` are independent of each other

Can't understand what is meant by this statement. `The oauth_skip_org_role_update_sync is independent  from a specific providers' skip_org_role_update_sync`?

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 25]
```diff
@@ -858,6 +860,21 @@ By default, the users' organization and role is reset on every new login.
...
+
+A table detailing what happens when the legacy option is enabled/disabled and an auth specific version is enabled/disabled is shown below.
+
+The table below shows the available OAuth providers and their setting with the default value and the skip org role sync setting.
+| OAuth Provider | `oauth_skip_org_role_sync_update` | `skip_org_role_sync` | Behavior |
+| --- | --- | --- | --- | --- |
+| Grafana.com | true | false | will still sync Org Roles from Grafana.Com |
```
Comment (Jguer) [inline docs/sources/setup-grafana/configure-grafana/_index.md:874]:
```suggestion
| Grafana.com | true | false | skip org role sync for grafana.com |
```

This should be the correct behavior otherwise it will break existing user's configs

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 26]
```diff
@@ -858,6 +860,21 @@ By default, the users' organization and role is reset on every new login.
...
+
+If you want to prevent org roles from being synced for a specific OAuth integration, set `skip_org_role_update_sync` to `true`.
+
+The `oauth_skip_org_role_update_sync` are independent of each other. `skip_org_role_sync` is specific to the provider integration, while `oauth_skip_org_role_sync_update` is a generic setting that only applies to `Github`, `Gitlab`, `AzureAD`, `Generic OAuth` integrations. We are deprecating the generic setting in favor of the provider specific settings.
+
+A table detailing what happens when the legacy option is enabled/disabled and an auth specific version is enabled/disabled is shown below.
+
+The table below shows the available OAuth providers and their setting with the default value and the skip org role sync setting.
+| OAuth Provider | `oauth_skip_org_role_sync_update` | `skip_org_role_sync` | Behavior |
+| --- | --- | --- | --- | --- |
```
Comment (Jguer) [inline docs/sources/setup-grafana/configure-grafana/_index.md:873]:
```suggestion
| --- | --- | --- | --- |
```

extra bar makes an invalid table in .md

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 27]
```diff
@@ -32,17 +32,15 @@ client_secret = eyJ0Ijoib2F1dGgyYyIhlmlkIjoiNzUwYmMzM2MxMGRjMjE6NDh3OWQiLCJ2Ijoi
...
 enabled = true
-
-# prevents the sync of org roles from Grafana.com
-skip_org_role_sync = false
 ```
 
 ## Skip organization role sync
 
-To prevent the sync of org roles from Grafana.com, set `skip_org_role_sync` to `true`. This is useful if you want to use Grafana.com as an identity provider but want to manage the org roles in Grafana.
+To prevent the sync of org roles from Grafana.com, set `skip_org_role_sync` to `true`. This is useful if you want to use your Grafana instance to manage the organization roles for your users.
```
Comment (Jguer) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:39]:
```suggestion
To prevent the sync of org roles from Grafana.com, set `skip_org_role_sync` to `true`. This is useful if you want to manage the organization roles for your users from within Grafana.
```

File: conf/sample.ini [hunk 28]
```diff
@@ -679,7 +680,6 @@
 # At 1 am every day
 ;sync_cron = "0 1 * * *"
 ;active_sync_enabled = true
-
```
Comment (alexanderzobnin) [inline conf/sample.ini:682]: I think there should be a blank line

**Review by alexanderzobnin — COMMENTED** [review]
Just a minor things like formatting

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 25 above)
Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-grafana/_index.md:874]: We do not have any Grafana.com for the "generic" setting. Meaning that the existing behavior is syncing the org roles for GrafanaCom, this is always why this is being prioritized.

If we want to change existing behavior for users then we will have to introduce GrafanaCom in the sync part as well.

---
Summarize:
- Do we want eventually that all of the OAuth integrations will skip the sync if the "generic" setting is set from the users; then we have to make this change and it will affect the behavior for current users.

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 24 above)
Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-grafana/_index.md:867]: Yea, they are not independant. Poor choice of words, I will reframe it more like separate settings

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 29]
```diff
@@ -858,6 +860,19 @@ By default, the users' organization and role is reset on every new login.
 > With Grafana 10, if `oauth_skip_org_role_update_sync` option is set to `false`, users with no mapping will be
 > reset to the default organization role on every login. [See `auto_assign_org_role` option]({{< relref ".#auto_assign_org_role" >}}).
 
+### [auth].<provider> skip_org_role_update_sync
+
+To prevent synchronization of organization roles for a specific OAuth integration, you can set the `skip_org_role_sync` option to `true`. Please note that there is also a separate setting called `oauth_skip_org_role_update_sync` which has a different scope. While `skip_org_role_sync` only applies to the provider integration, `oauth_skip_org_role_update_sync` is a generic setting that affects GitHub, GitLab, AzureAD, and Generic OAuth integrations. However, we are planning to phase out the use of the generic setting in favor of provider-specific settings.
```
Comment (eleijonmarck) [inline docs/sources/setup-grafana/configure-grafana/_index.md:865]: @Jguer your previews comment was removed infavor of the new doc string instead.

Is it possible to review this text, you mentioned that `independant` was incorrect to explain the differences between these settings.

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 30]
```diff
@@ -858,6 +860,19 @@ By default, the users' organization and role is reset on every new login.
...
+
+A table detailing what happens when the soon to be legacy option `oauth_skip_org_role_update_sync` is enabled/disabled and an auth specific setting `skip_org_role_sync` is enabled/disabled is shown below.
+
+The table below shows the available OAuth providers and their setting with the default value and the skip org role sync setting.
+| OAuth Provider | `oauth_skip_org_role_sync_update` | `skip_org_role_sync` | Behavior |
+| --- | --- | --- | --- |
+| Grafana.com | true | false | will still sync Org Roles from Grafana.Com |
+| Grafana.com | false | true | skip org role sync for grafana.com |
+| Grafana.com | true | true | no sync from any OAuth integration as well as Grafana.Com |
```
Comment (IevaVasiljeva) [inline docs/sources/setup-grafana/configure-grafana/_index.md:874]: The table is missing one a row where both options are `false`.

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 31]
```diff
@@ -0,0 +1,46 @@
+---
```
Comment (IevaVasiljeva) [inline docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:1]: I didn't realise that we didn't have a section for GCom. Thanks for adding it!

File: pkg/setting/setting.go [hunk 32]
```diff
@@ -456,12 +456,14 @@ type Cfg struct {
 	// then Live uses AppURL as the only allowed origin.
 	LiveAllowedOrigins []string
 
+	// Grafana.com
```
Comment (IevaVasiljeva) [inline pkg/setting/setting.go:459]: nit: seems like an unneeded change.

Comment (IevaVasiljeva) [inline pkg/setting/setting.go:459]: nit: seems like an unneeded change.

File: pkg/setting/setting.go [hunk 33]
```diff
@@ -456,12 +456,14 @@ type Cfg struct {
...
 	// Grafana.com URL, used for OAuth redirect.
 	GrafanaComURL string
 	// Grafana.com API URL. Can be set separately to GrafanaComURL
 	// in case API is not publicly accessible.
 	// Defaults to GrafanaComURL setting + "/api" if unset.
-	GrafanaComAPIURL string
+	GrafanaComAPIURL          string
+	GrafanaComSkipOrgRoleSync bool
```
Comment (IevaVasiljeva) [inline pkg/setting/setting.go:466]: nit: would be good to add a little comment for this option.

File: pkg/setting/setting.go [hunk 34]
```diff
@@ -1351,6 +1353,11 @@ func readSecuritySettings(iniFile *ini.File, cfg *Cfg) error {
 	return nil
 }
 
+func readAuthGrafanaComSettings(iniFile *ini.File, cfg *Cfg) {
+	sec := iniFile.Section("auth.grafana_com")
+	cfg.GrafanaComSkipOrgRoleSync = sec.Key("skip_org_role_sync").MustBool(false)
+}
```
Comment (IevaVasiljeva) [inline pkg/setting/setting.go:1359]: nit: not necessarily for this PR, but it would be good to have a follow-up refactor to read all Grafana Com settings here.

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 25 above)
Comment (Jguer) [reply docs/sources/setup-grafana/configure-grafana/_index.md:872]: Maybe the frontend is not taking it into account but the backend is

https://github.com/grafana/grafana/blob/main/pkg/api/login_oauth.go#L288 should affect all oauth integrations

 We had to change this setting for ops.grafana because we enforced the sync for a while.

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 29 above)
Comment (Jguer) [reply docs/sources/setup-grafana/configure-grafana/_index.md:865]:
```suggestion
To prevent synchronization of organization roles for a specific OAuth integration, you can set the `skip_org_role_sync` option to `true`. Please note that there is also a separate setting called `oauth_skip_org_role_update_sync` which has a different scope. While `skip_org_role_sync` only applies to the specific OAuth provider, `oauth_skip_org_role_update_sync` is a generic setting that affects all configured OAuth providers. 

 The setting `oauth_skip_org_role_update_sync` is being deprecated in favor of provider-specific settings.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 25 above)
Comment (IevaVasiljeva) [reply docs/sources/setup-grafana/configure-grafana/_index.md:872]: I don't really follow your comment Eric, but I think what Joao suggested is the behaviour that we want and is also the behaviour that we have after the changes in this PR - org role sync for GCom will be disabled if either the generic `oauth_skip_org_role_sync_update` is set to true or if `skip_org_role_sync` for GCom is set to true.

This is where the generic setting is looked at when setting org roles for the external user:
https://github.com/grafana/grafana/blob/main/pkg/api/login_oauth.go#L288

Note that I haven't tested this myself, but I think this is the behaviour that we saw when we were testing together Eric?

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 23 above)
Comment (IevaVasiljeva) [reply docs/sources/setup-grafana/configure-grafana/_index.md:863]: I would probably prefer to have separate explicit sections for each provider (ie, this one would be [auth.grafana_com] skip_org_role_update_sync. That would also make it very clear for which providers we have a provider specific setting and for which we don't have it yet.

**Review by IevaVasiljeva — COMMENTED** [review]
Thanks for adding the docs and the test! I think the PR is almost there, I just have some suggestions for doc changes.

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 25 above)
Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-grafana/_index.md:874]: before I made my comment I tested this out, let me get back to you with a test to see that was true. Yeas to me I login and the setting `oauth_skip_org_role_sync_update` is set to `true`.


https://user-images.githubusercontent.com/7727602/211037777-3420ccd6-c72e-426d-9340-d7ea3a6f88ff.mov

Does this make sense to you? That we are unable to set the org roles "basic,viewer and admin"

Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-grafana/_index.md:874]: when we have the other setting enabled we are able to edit the org roles.

here the setting `skip_org_role_sync` in `[auth].grafanacom` is set to `true`.

https://user-images.githubusercontent.com/7727602/211038725-c2a5f15c-5977-4aaf-998f-7f9e6304b851.mov

Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-grafana/_index.md:874]: if we want it to be that `oauth_skip_org_role_sync_update` is set to `true` would enable us to also edit the org roles. Then we would need to change this PR to reflect that

Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-grafana/_index.md:874]: hey! i am realizing i am only checking the frontend settings here and we are talking about what the function signature for the externalUser is sending. Let me revisit this

Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-grafana/_index.md:874]: > Maybe the frontend is not taking it into account but the backend is

summary:
- backend is taking it into account but not the frontend

changing the frontend to make it reflect what want as desired behaviour

File: docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md [hunk 35]
```diff
@@ -0,0 +1,46 @@
...
+aliases:
+  - ../../../auth/grafana-com/
+description: Grafana Com Authentication
+title: Configure Grafana Com authentication
+weight: 500
+---
+
+# Configure Grafana Com authentication
+
+To enable GrafanaCom as your authentication provider, you configure it to generate a client ID and a secret key.
```
[the remark this answers is not in the document]
Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-security/configure-authentication/grafana-com/index.md:11]: yea, i asked in the GrafanaCom channel and there does not seem to be a standard way as far as I know

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 25 above)
Comment (Jguer) [reply docs/sources/setup-grafana/configure-grafana/_index.md:874]: :heavy_check_mark:  changing the frontend to make it reflect what want as desired behaviour

:heavy_check_mark:  desired behavior being - we want the "generic" setting to make it so that all OAuth integrations are not synced when set. And when a specific provider is set, only those users should not be synced

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 23 above)
Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-grafana/_index.md:863]: let's do one by one first. Thought it might make it a lot of documentation and maybe each provider might be the same settings to reflect. Who knows?

Let's start with one by one!

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 36]
```diff
@@ -858,6 +860,20 @@ By default, the users' organization and role is reset on every new login.
 > With Grafana 10, if `oauth_skip_org_role_update_sync` option is set to `false`, users with no mapping will be
 > reset to the default organization role on every login. [See `auto_assign_org_role` option]({{< relref ".#auto_assign_org_role" >}}).
 
+### [auth].grafanacom skip_org_role_update_sync
+
+To prevent synchronization of organization roles for a specific OAuth integration, you can set the `skip_org_role_sync` option to `true`. Please note that there is also a separate setting called `oauth_skip_org_role_update_sync` which has a different scope. While `skip_org_role_sync` only applies to the specific OAuth provider, `oauth_skip_org_role_update_sync` is a generic setting that affects all configured OAuth providers.
+
+The setting `oauth_skip_org_role_update_sync` is deprecated in favor of provider-specific settings.
```
Comment (IevaVasiljeva) [inline docs/sources/setup-grafana/configure-grafana/_index.md:867]:
```suggestion
The setting `oauth_skip_org_role_update_sync` will be deprecated in favor of provider-specific settings.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 37]
```diff
@@ -858,6 +860,20 @@ By default, the users' organization and role is reset on every new login.
...
+
+The table below shows the available OAuth providers and their setting with the default value and the skip org role sync setting.
+| OAuth Provider | `oauth_skip_org_role_sync_update` | `skip_org_role_sync` | Behavior |
+| --- | --- | --- | --- |
+| Grafana.com | false | false | will sync with Grafana.Com roles |
```
Comment (IevaVasiljeva) [inline docs/sources/setup-grafana/configure-grafana/_index.md:872]:
```suggestion
| Grafana.com | false | false | will sync with Grafana.com roles |
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 38]
```diff
@@ -858,6 +860,20 @@ By default, the users' organization and role is reset on every new login.
...
+| Grafana.com | true | false | skip org role sync for OAuth providers including grafana.com users |
```
Comment (IevaVasiljeva) [inline docs/sources/setup-grafana/configure-grafana/_index.md:873]:
```suggestion
| Grafana.com | true | false | skip org role sync for OAuth providers including Grafana.com users |
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 39]
```diff
@@ -858,6 +860,20 @@ By default, the users' organization and role is reset on every new login.
...
+| Grafana.com | false | true | skip org role sync for grafana.com users |
```
Comment (IevaVasiljeva) [inline docs/sources/setup-grafana/configure-grafana/_index.md:874]:
```suggestion
| Grafana.com | false | true | skip org role sync for Grafana.com users, but not for other OAuth providers |
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 40]
```diff
@@ -858,6 +860,20 @@ By default, the users' organization and role is reset on every new login.
...
+| Grafana.com | true | true | skip org role sync for grafana.com users in addition all other OAuth providers |
```
Comment (IevaVasiljeva) [inline docs/sources/setup-grafana/configure-grafana/_index.md:875]:
```suggestion
| Grafana.com | true | true | skip org role sync for Grafana.com users and all other OAuth providers |
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 41]
```diff
@@ -858,6 +860,20 @@ By default, the users' organization and role is reset on every new login.
 > With Grafana 10, if `oauth_skip_org_role_update_sync` option is set to `false`, users with no mapping will be
 > reset to the default organization role on every login. [See `auto_assign_org_role` option]({{< relref ".#auto_assign_org_role" >}}).
 
+### [auth].grafanacom skip_org_role_update_sync
```
Comment (IevaVasiljeva) [inline docs/sources/setup-grafana/configure-grafana/_index.md:863]:
```suggestion
### [auth.grafana_com] skip_org_role_update_sync
```

File: pkg/setting/setting.go (hunk 34 above)
Comment (IevaVasiljeva) [reply pkg/setting/setting.go:1361]: Yup, exactly :)

**Review by IevaVasiljeva — APPROVED** [review]
Nice work! I've added some minor suggestions for docs, but apart from that I think this is ready to be shipped 🚢 

Thanks for adding the demo and also writing the tests!

**Comment by gabor** [comment]
(sorry, just wanted to remove `grafana-observability-logs` from the list of reviewers because this does not seem related to our squad. but github somehow removed a bunch of other squads too, pls add back those that you really need. sorry)

**Comment by eleijonmarck** [comment]
> (sorry, just wanted to remove `grafana-observability-logs` from the list of reviewers because this does not seem related to our squad. but github somehow removed a bunch of other squads too, pls add back those that you really need. sorry)

this is a merge gone wrong! Hope it is understandable for all the team 🙏  sorry
